File: android/src/main/kotlin/com/khanhtq/app_blocker/AppBlockerPlugin.kt
package com.khanhtq.app_blocker

import android.app.Activity
import android.content.Context
import com.khanhtq.app_blocker.blocking.BlockingServiceManager
import com.khanhtq.app_blocker.event.BlockEventStreamHandler
import com.khanhtq.app_blocker.persistence.BlockerPreferences
import com.khanhtq.app_blocker.scheduling.ProfileManager
import com.khanhtq.app_blocker.scheduling.ScheduleManager
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

/**
 * Flutter plugin entry-point for app_blocker.
 *
 * Wires the Dart method/event channels to the native Android managers.
 * All method calls that may block are dispatched to [Dispatchers.Default]
 * and the result is posted back on the main thread.
 */
class AppBlockerPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {

    companion object {
        // Channel names — must match Dart AppBlockerConstants exactly.
        private const val METHOD_CHANNEL = "com.khanhtq.app_blocker/methods"
        private const val EVENT_CHANNEL = "com.khanhtq.app_blocker/events"

        // Method names.
        private const val GET_CAPABILITIES = "getCapabilities"
        private const val CHECK_PERMISSION = "checkPermission"
        private const val REQUEST_PERMISSION = "requestPermission"
        private const val GET_APPS = "getApps"
        private const val BLOCK_APPS = "blockApps"
        private const val BLOCK_ALL = "blockAll"
        private const val UNBLOCK_APPS = "unblockApps"
        private const val UNBLOCK_ALL = "unblockAll"
        private const val GET_BLOCKED_APPS = "getBlockedApps"
        private const val GET_APP_STATUS = "getAppStatus"
        private const val SET_BLOCK_SCREEN_CONFIG = "setBlockScreenConfig"
        private const val GET_BLOCK_SCREEN_CONFIG = "getBlockScreenConfig"
        private const val ADD_SCHEDULE = "addSchedule"
        private const val UPDATE_SCHEDULE = "updateSchedule"
        private const val REMOVE_SCHEDULE = "removeSchedule"
        private const val GET_SCHEDULES = "getSchedules"
        private const val ENABLE_SCHEDULE = "enableSchedule"
        private const val DISABLE_SCHEDULE = "disableSchedule"
        private const val CREATE_PROFILE = "createProfile"
        private const val UPDATE_PROFILE = "updateProfile"
        private const val DELETE_PROFILE = "deleteProfile"
        private const val GET_PROFILES = "getProfiles"
        private const val ACTIVATE_PROFILE = "activateProfile"
        private const val DEACTIVATE_PROFILE = "deactivateProfile"
        private const val GET_ACTIVE_PROFILE = "getActiveProfile"

        // Error codes — must match Dart AppBlockerConstants exactly.
        const val ERROR_PERMISSION_DENIED = "PERMISSION_DENIED"
        const val ERROR_SERVICE_UNAVAILABLE = "SERVICE_UNAVAILABLE"
        const val ERROR_SCHEDULE_CONFLICT = "SCHEDULE_CONFLICT"
        const val ERROR_PROFILE_NOT_FOUND = "PROFILE_NOT_FOUND"
        const val ERROR_INVALID_CONFIG = "INVALID_CONFIG"
    }

    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private lateinit var context: Context
    private var activity: Activity? = null

    private lateinit var permissionManager: PermissionManager
    private lateinit var appResolver: AppResolver
    private lateinit var blockingServiceManager: BlockingServiceManager
    private lateinit var scheduleManager: ScheduleManager
    private lateinit var profileManager: ProfileManager
    private lateinit var preferences: BlockerPreferences

    // Coroutine scope bound to the plugin lifecycle.
    private val job = Job()
    private val scope = CoroutineScope(Dispatchers.Default + job)

    // ------------------------------------------------------------------
    // FlutterPlugin
    // ------------------------------------------------------------------

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        preferences = BlockerPreferences(context)

        methodChannel = MethodChannel(binding.binaryMessenger, METHOD_CHANNEL)
        methodChannel.setMethodCallHandler(this)

        eventChannel = EventChannel(binding.binaryMessenger, EVENT_CHANNEL)
        eventChannel.setStreamHandler(BlockEventStreamHandler.instance)

        permissionManager = PermissionManager(context)
        appResolver = AppResolver(context)
        blockingServiceManager = BlockingServiceManager(context)
        scheduleManager = ScheduleManager(context)
        profileManager = ProfileManager(context)
        scheduleManager.rescheduleAll()
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        job.cancel()
    }

    // ------------------------------------------------------------------
    // ActivityAware
    // ------------------------------------------------------------------

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        activity = null
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    // ------------------------------------------------------------------
    // MethodCallHandler
    // ------------------------------------------------------------------

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {

            // -- Capabilities --

            GET_CAPABILITIES -> {
                result.success(
                    mapOf(
                        "canBlockApps" to true,
                        "canCustomizeBlockScreen" to true,
                        "canUseSystemShield" to false,
                        "canSchedule" to true,
                        "canGetInstalledApps" to true,
                        "canShowActivityPicker" to false,
                    )
                )
            }

            // -- Permissions --

            CHECK_PERMISSION -> result.success(permissionManager.checkAllPermissions())

            REQUEST_PERMISSION -> {
                val currentActivity = activity ?: run {
                    result.error(ERROR_SERVICE_UNAVAILABLE, "Activity is not available.", null)
                    return
                }
                result.success(permissionManager.requestAllPermissions(currentActivity))
            }

            // -- App Discovery --

            GET_APPS -> {
                scope.launch {
                    try {
                        val apps = appResolver.getInstalledApps()
                        withContext(Dispatchers.Main) { result.success(apps) }
                    } catch (e: Exception) {
                        withContext(Dispatchers.Main) {
                            result.error(ERROR_SERVICE_UNAVAILABLE, e.message, null)
                        }
                    }
                }
            }

            // -- Blocking --

            BLOCK_APPS -> {
                val identifiers = call.argument<List<String>>("identifiers") ?: run {
                    result.error(ERROR_INVALID_CONFIG, "Missing 'identifiers' argument.", null)
                    return
                }
                try {
                    blockingServiceManager.startBlocking(identifiers)
                    result.success(null)
                } catch (e: Exception) {
                    result.error(ERROR_SERVICE_UNAVAILABLE, e.message, null)
                }
            }

            BLOCK_ALL -> {
                try {
                    blockingServiceManager.startBlockingAll()
                    result.success(null)
                } catch (e: Exception) {
                    result.error(ERROR_SERVICE_UNAVAILABLE, e.message, null)
                }
            }

            UNBLOCK_APPS -> {
                val identifiers = call.argument<List<String>>("identifiers") ?: run {
                    result.error(ERROR_INVALID_CONFIG, "Missing 'identifiers' argument.", null)
                    return
                }
                blockingServiceManager.stopBlockingApps(identifiers)
                result.success(null)
            }

            UNBLOCK_ALL -> {
                profileManager.deactivateActiveProfile()
                scheduleManager.disableAll()
                blockingServiceManager.stopBlocking()
                result.success(null)
            }

            GET_BLOCKED_APPS -> {
                if (preferences.isBlocking() && preferences.isBlockAll()) {
                    result.success(listOf("__all__"))
                } else {
                    result.success(
                        (blockingServiceManager.getBlockedApps() + scheduleManager.getActivelyBlockedApps()).toList()
                    )
                }
            }

            GET_APP_STATUS -> {
                val identifier = call.argument<String>("identifier") ?: run {
                    result.error(ERROR_INVALID_CONFIG, "Missing 'identifier' argument.", null)
                    return
                }
                val blocked = blockingServiceManager.getBlockedApps() + scheduleManager.getActivelyBlockedApps()
                val status = when {
                    preferences.isBlocking() && preferences.isBlockAll() -> "blocked"
                    identifier in blocked -> "blocked"
                    else -> "unblocked"
                }
                result.success(status)
            }

            // -- Block Screen Config --

            SET_BLOCK_SCREEN_CONFIG -> {
                val configMap = buildMap<String, Any?> {
                    put("title", call.argument<String>("title"))
                    put("subtitle", call.argument<String>("subtitle"))
                    put("message", call.argument<String>("message"))
                    put("backgroundColor", call.argument<Long>("backgroundColor"))
                    put("iconAssetPath", call.argument<String>("iconAssetPath"))
                }
                preferences.overlayConfig = com.google.gson.Gson().toJson(configMap)
                result.success(null)
            }

            GET_BLOCK_SCREEN_CONFIG -> {
                val json = preferences.overlayConfig
                if (json == "{}") {
                    result.success(null)
                } else {
                    try {
                        val obj = org.json.JSONObject(json)
                        val map = mutableMapOf<String, Any?>()
                        if (obj.has("title") && !obj.isNull("title")) map["title"] = obj.getString("title")
                        if (obj.has("subtitle") && !obj.isNull("subtitle")) map["subtitle"] = obj.getString("subtitle")
                        if (obj.has("message") && !obj.isNull("message")) map["message"] = obj.getString("message")
                        if (obj.has("backgroundColor") && !obj.isNull("backgroundColor")) map["backgroundColor"] = obj.getLong("backgroundColor")
                        if (obj.has("iconAssetPath") && !obj.isNull("iconAssetPath")) map["iconAssetPath"] = obj.getString("iconAssetPath")
                        result.success(map)
                    } catch (_: Exception) {
                        result.success(null)
                    }
                }
            }

            // -- Scheduling --

            ADD_SCHEDULE -> {
                // Method channel guarantees Map<String, Any?> for object arguments.
                val scheduleMap = requireMapArgument(call, result) ?: return
                try {
                    scheduleManager.addSchedule(scheduleMap)
                    result.success(null)
                } catch (e: Exception) {
                    result.error(ERROR_SCHEDULE_CONFLICT, e.message, null)
                }
            }

            UPDATE_SCHEDULE -> {
                val scheduleMap = requireMapArgument(call, result) ?: return
                try {
                    scheduleManager.updateSchedule(scheduleMap)
                    result.success(null)
                } catch (e: Exception) {
                    result.error(ERROR_SCHEDULE_CONFLICT, e.message, null)
                }
            }

            REMOVE_SCHEDULE -> {
                val id = call.argument<String>("id") ?: run {
                    result.error(ERROR_INVALID_CONFIG, "Missing 'id' argument.", null)
                    return
                }
                scheduleManager.removeSchedule(id)
                result.success(null)
            }

            GET_SCHEDULES -> result.success(scheduleManager.getSchedules())

            ENABLE_SCHEDULE -> {
                val id = call.argument<String>("id") ?: run {
                    result.error(ERROR_INVALID_CONFIG, "Missing 'id' argument.", null)
                    return
                }
                scheduleManager.enableSchedule(id)
                result.success(null)
            }

            DISABLE_SCHEDULE -> {
                val id = call.argument<String>("id") ?: run {
                    result.error(ERROR_INVALID_CONFIG, "Missing 'id' argument.", null)
                    return
                }
                scheduleManager.disableSchedule(id)
                result.success(null)
            }

            // -- Profiles --

            CREATE_PROFILE -> {
                val profileMap = requireMapArgument(call, result) ?: return
                try {
                    profileManager.createProfile(profileMap)
                    result.success(null)
                } catch (e: Exception) {
                    result.error(ERROR_INVALID_CONFIG, e.message, null)
                }
            }

            UPDATE_PROFILE -> {
                val profileMap = requireMapArgument(call, result) ?: return
                try {
                    profileManager.updateProfile(profileMap)
                    result.success(null)
                } catch (e: Exception) {
                    result.error(ERROR_PROFILE_NOT_FOUND, e.message, null)
                }
            }

            DELETE_PROFILE -> {
                val id = call.argument<String>("id") ?: run {
                    result.error(ERROR_INVALID_CONFIG, "Missing 'id' argument.", null)
                    return
                }
                profileManager.deleteProfile(id)
                result.success(null)
            }

            GET_PROFILES -> result.success(profileManager.getProfiles())

            ACTIVATE_PROFILE -> {
                val id = call.argument<String>("id") ?: run {
                    result.error(ERROR_INVALID_CONFIG, "Missing 'id' argument.", null)
                    return
                }
                try {
                    profileManager.activateProfile(id)
                    result.success(null)
                } catch (e: Exception) {
                    result.error(ERROR_PROFILE_NOT_FOUND, e.message, null)
                }
            }

            DEACTIVATE_PROFILE -> {
                val id = call.argument<String>("id") ?: run {
                    result.error(ERROR_INVALID_CONFIG, "Missing 'id' argument.", null)
                    return
                }
                profileManager.deactivateProfile(id)
                result.success(null)
            }

            GET_ACTIVE_PROFILE -> result.success(profileManager.getActiveProfile())

            else -> result.notImplemented()
        }
    }

    // ------------------------------------------------------------------
    // Helpers
    // ------------------------------------------------------------------

    /**
     * Extracts [MethodCall.arguments] as `Map<String, Any?>`, sending an
     * [ERROR_INVALID_CONFIG] error and returning `null` if the arguments
     * are not a map.
     *
     * Flutter's method channel guarantees that object arguments arrive as
     * `Map<String, Any?>` when sent from Dart as a `Map<String, dynamic>`.
     * The unchecked cast is therefore safe; the suppress is intentional.
     */
    private fun requireMapArgument(call: MethodCall, result: Result): Map<String, Any?>? {
        val raw = call.arguments
        if (raw !is Map<*, *>) {
            result.error(ERROR_INVALID_CONFIG, "Expected a map argument for '${call.method}'.", null)
            return null
        }
        @Suppress("UNCHECKED_CAST") // Safe: Flutter channel always uses String keys with Any? values
        return raw as Map<String, Any?>
    }
}

File: ios/Classes/AppBlockerPlugin.swift
import Flutter
import UIKit

/// Flutter plugin entry-point for app_blocker on iOS.
public class AppBlockerPlugin: NSObject, FlutterPlugin {

    private var permissionManager: PermissionManager?
    var shieldManager: ShieldManager?
    private var activityPickerCoordinator: ActivityPickerCoordinator?
    private var scheduleManager: ScheduleManager?
    private var profileManager: ProfileManager?
    private var eventStreamHandler: BlockEventStreamHandler?

    /// Shared instance exposed to `ActivityPickerCoordinator` for callbacks.
    static var shared: AppBlockerPlugin?

    // MARK: - Registration

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "com.khanhtq.app_blocker/methods",
            binaryMessenger: registrar.messenger()
        )
        let eventChannel = FlutterEventChannel(
            name: "com.khanhtq.app_blocker/events",
            binaryMessenger: registrar.messenger()
        )

        let instance = AppBlockerPlugin()
        AppBlockerPlugin.shared = instance

        let shield = ShieldManager()
        instance.shieldManager = shield
        instance.scheduleManager = ScheduleManager()
        instance.profileManager = ProfileManager(shieldManager: shield)
        instance.permissionManager = PermissionManager()
        instance.activityPickerCoordinator = ActivityPickerCoordinator()

        instance.eventStreamHandler = BlockEventStreamHandler()
        eventChannel.setStreamHandler(instance.eventStreamHandler)

        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    // MARK: - Method dispatch

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getCapabilities":     handleGetCapabilities(result: result)
        case "checkPermission":     handleCheckPermission(result: result)
        case "requestPermission":   handleRequestPermission(result: result)
        case "getApps":             handleGetApps(result: result)
        case "blockApps":           handleBlockApps(call: call, result: result)
        case "blockAll":            handleBlockAll(result: result)
        case "unblockApps":         handleUnblockApps(call: call, result: result)
        case "unblockAll":          handleUnblockAll(result: result)
        case "getBlockedApps":      handleGetBlockedApps(result: result)
        case "getAppStatus":        handleGetAppStatus(call: call, result: result)
        case "setBlockScreenConfig":
            result(FlutterError(
                code: "PLATFORM_UNSUPPORTED",
                message: "Block screen configuration is not supported on iOS.",
                details: nil
            ))
        case "addSchedule":         handleAddSchedule(call: call, result: result)
        case "updateSchedule":      handleUpdateSchedule(call: call, result: result)
        case "removeSchedule":      handleRemoveSchedule(call: call, result: result)
        case "getSchedules":        handleGetSchedules(result: result)
        case "enableSchedule":      handleEnableSchedule(call: call, result: result)
        case "disableSchedule":     handleDisableSchedule(call: call, result: result)
        case "createProfile":       handleCreateProfile(call: call, result: result)
        case "updateProfile":       handleUpdateProfile(call: call, result: result)
        case "deleteProfile":       handleDeleteProfile(call: call, result: result)
        case "getProfiles":         handleGetProfiles(result: result)
        case "activateProfile":     handleActivateProfile(call: call, result: result)
        case "deactivateProfile":   handleDeactivateProfile(call: call, result: result)
        case "getActiveProfile":    handleGetActiveProfile(result: result)
        default:                    result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Capabilities

    private func handleGetCapabilities(result: @escaping FlutterResult) {
        result([
            "canBlockApps": true,
            "canCustomizeBlockScreen": false,
            "canUseSystemShield": true,
            "canSchedule": false,
            "canGetInstalledApps": false,
            "canShowActivityPicker": true,
        ] as [String: Any])
    }

    // MARK: - Permissions

    private func handleCheckPermission(result: @escaping FlutterResult) {
        guard let manager = permissionManager else {
            return result(unavailableError("Permission manager"))
        }
        result(manager.checkPermission())
    }

    private func handleRequestPermission(result: @escaping FlutterResult) {
        guard let manager = permissionManager else {
            return result(unavailableError("Permission manager"))
        }
        manager.requestPermission(result: result)
    }

    // MARK: - App Discovery

    private func handleGetApps(result: @escaping FlutterResult) {
        guard let coordinator = activityPickerCoordinator else {
            return result(unavailableError("Activity picker coordinator"))
        }
        guard let rootVC = Self.findRootViewController() else {
            return result(unavailableError("Root view controller"))
        }
        coordinator.showPicker(from: rootVC, result: result)
    }

    // MARK: - Blocking

    private func handleBlockApps(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let shield = shieldManager else {
            return result(unavailableError("Shield manager"))
        }
        guard let args = call.arguments as? [String: Any],
              let identifiers = args["identifiers"] as? [String] else {
            return result(invalidConfigError("Missing 'identifiers' argument."))
        }
        shield.blockApps(identifiers: identifiers)
        eventStreamHandler?.sendEvent(type: "blocked", packageName: identifiers.first, scheduleId: nil)
        result(nil)
    }

    private func handleBlockAll(result: @escaping FlutterResult) {
        guard let shield = shieldManager else {
            return result(unavailableError("Shield manager"))
        }
        shield.blockAll()
        eventStreamHandler?.sendEvent(type: "blocked", packageName: nil, scheduleId: nil)
        result(nil)
    }

    private func handleUnblockApps(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let shield = shieldManager else {
            return result(unavailableError("Shield manager"))
        }
        guard let args = call.arguments as? [String: Any],
              let identifiers = args["identifiers"] as? [String] else {
            return result(invalidConfigError("Missing 'identifiers' argument."))
        }
        shield.unblockApps(identifiers: identifiers)
        eventStreamHandler?.sendEvent(type: "unblocked", packageName: identifiers.first, scheduleId: nil)
        result(nil)
    }

    private func handleUnblockAll(result: @escaping FlutterResult) {
        guard let shield = shieldManager else {
            return result(unavailableError("Shield manager"))
        }
        if let manager = profileManager,
           let activeProfile = manager.getActiveProfile(),
           let profileId = activeProfile["id"] as? String {
            manager.deactivateProfile(id: profileId)
            emitProfileEvents(profileId: profileId, appIdentifiers: activeProfile["appIdentifiers"] as? [String] ?? [], activated: false)
        }
        shield.unblockAll()
        eventStreamHandler?.sendEvent(type: "unblocked", packageName: nil, scheduleId: nil)
        result(nil)
    }

    private func emitProfileEvents(profileId: String, appIdentifiers: [String], activated: Bool) {
        for identifier in appIdentifiers {
            eventStreamHandler?.sendEvent(type: activated ? "blocked" : "unblocked", packageName: identifier, scheduleId: nil, profileId: profileId)
        }
        eventStreamHandler?.sendEvent(type: activated ? "profileActivated" : "profileDeactivated", packageName: nil, scheduleId: nil, profileId: profileId)
    }

    private func handleGetBlockedApps(result: @escaping FlutterResult) {
        guard let shield = shieldManager else {
            return result(unavailableError("Shield manager"))
        }
        result(shield.getBlockedApps())
    }

    private func handleGetAppStatus(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let shield = shieldManager else {
            return result(unavailableError("Shield manager"))
        }
        guard let args = call.arguments as? [String: Any],
              let identifier = args["identifier"] as? String else {
            return result(invalidConfigError("Missing 'identifier' argument."))
        }
        let blocked = shield.getBlockedApps()
        if blocked.contains(identifier) {
            result("blocked")
        } else if let schedMgr = scheduleManager {
            let isScheduled = schedMgr.getSchedules().contains { schedule in
                (schedule["enabled"] as? Bool == true) &&
                (schedule["appIdentifiers"] as? [String] ?? []).contains(identifier)
            }
            result(isScheduled ? "scheduled" : "unblocked")
        } else {
            result("unblocked")
        }
    }

    // MARK: - Scheduling (unsupported on iOS)
    // Schedule enforcement requires time-based background wakeups that are not
    // available without the DeviceActivity framework. All schedule methods return
    // PLATFORM_UNSUPPORTED. Check canSchedule via getCapabilities() before use.

    private func schedulingUnsupportedError() -> FlutterError {
        FlutterError(
            code: "PLATFORM_UNSUPPORTED",
            message: "Schedule-based blocking is not supported on iOS.",
            details: nil
        )
    }

    private func handleAddSchedule(call: FlutterMethodCall, result: @escaping FlutterResult) {
        result(schedulingUnsupportedError())
    }

    private func handleUpdateSchedule(call: FlutterMethodCall, result: @escaping FlutterResult) {
        result(schedulingUnsupportedError())
    }

    private func handleRemoveSchedule(call: FlutterMethodCall, result: @escaping FlutterResult) {
        result(schedulingUnsupportedError())
    }

    private func handleGetSchedules(result: @escaping FlutterResult) {
        result(schedulingUnsupportedError())
    }

    private func handleEnableSchedule(call: FlutterMethodCall, result: @escaping FlutterResult) {
        result(schedulingUnsupportedError())
    }

    private func handleDisableSchedule(call: FlutterMethodCall, result: @escaping FlutterResult) {
        result(schedulingUnsupportedError())
    }

    // MARK: - Profiles

    private func handleCreateProfile(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let manager = profileManager else {
            return result(unavailableError("Profile manager"))
        }
        guard let data = call.arguments as? [String: Any] else {
            return result(invalidConfigError("Invalid profile data."))
        }
        manager.createProfile(data: data)
        result(nil)
    }

    private func handleUpdateProfile(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let manager = profileManager else {
            return result(unavailableError("Profile manager"))
        }
        guard let data = call.arguments as? [String: Any] else {
            return result(invalidConfigError("Invalid profile data."))
        }
        manager.updateProfile(data: data)
        result(nil)
    }

    private func handleDeleteProfile(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let manager = profileManager else {
            return result(unavailableError("Profile manager"))
        }
        guard let args = call.arguments as? [String: Any],
              let id = args["id"] as? String else {
            return result(invalidConfigError("Missing 'id' argument."))
        }
        manager.deleteProfile(id: id)
        result(nil)
    }

    private func handleGetProfiles(result: @escaping FlutterResult) {
        guard let manager = profileManager else {
            return result(unavailableError("Profile manager"))
        }
        result(manager.getProfiles())
    }

    private func handleActivateProfile(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let manager = profileManager else {
            return result(unavailableError("Profile manager"))
        }
        guard let args = call.arguments as? [String: Any],
              let id = args["id"] as? String else {
            return result(invalidConfigError("Missing 'id' argument."))
        }
        let appIdentifiers = (manager.getProfiles()
            .first { ($0["id"] as? String) == id }?["appIdentifiers"] as? [String]) ?? []
        if manager.activateProfile(id: id) {
            emitProfileEvents(profileId: id, appIdentifiers: appIdentifiers, activated: true)
            result(nil)
        } else {
            result(FlutterError(
                code: "PROFILE_NOT_FOUND",
                message: "Profile '\(id)' not found.",
                details: nil
            ))
        }
    }

    private func handleDeactivateProfile(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let manager = profileManager else {
            return result(unavailableError("Profile manager"))
        }
        guard let args = call.arguments as? [String: Any],
              let id = args["id"] as? String else {
            return result(invalidConfigError("Missing 'id' argument."))
        }
        let appIdentifiers = (manager.getProfiles()
            .first { ($0["id"] as? String) == id }?["appIdentifiers"] as? [String]) ?? []
        manager.deactivateProfile(id: id)
        emitProfileEvents(profileId: id, appIdentifiers: appIdentifiers, activated: false)
        result(nil)
    }

    private func handleGetActiveProfile(result: @escaping FlutterResult) {
        guard let manager = profileManager else {
            return result(unavailableError("Profile manager"))
        }
        result(manager.getActiveProfile())
    }

    // MARK: - Error helpers

    private func unavailableError(_ component: String) -> FlutterError {
        FlutterError(
            code: "SERVICE_UNAVAILABLE",
            message: "\(component) is not available.",
            details: nil
        )
    }

    private func invalidConfigError(_ message: String) -> FlutterError {
        FlutterError(code: "INVALID_CONFIG", message: message, details: nil)
    }

    // MARK: - Utilities

    private static func findRootViewController() -> UIViewController? {
        return UIApplication.shared.connectedScenes
            .compactMap { $0 as? UIWindowScene }
            .flatMap { $0.windows }
            .first { $0.isKeyWindow }?
            .rootViewController
    }
}
